src/serac/numerics/petsc_solvers.hpp

```diff
@@ -0,0 +1,516 @@
+   * Use `-pc_type TYPE -help` for a complete list of options.
+   * @note If @a prefix is provided, the options should be set as `-[prefix]_[option] [value]`.
+   */
+  PetscPCSolver(MPI_Comm comm, mfem::Operator& op, PCType pc_type = PCJACOBI,
+                const std::string& prefix = std::string());
+
+  virtual void SetOperator(const Operator& op);
+  virtual void Mult(const mfem::Vector& b, mfem::Vector& x) const;
+  virtual void MultTranspose(const mfem::Vector& b, mfem::Vector& x) const;
+};
+
+/**
+ * @brief A PETSC-based preconditioner which requires information about the underlying finite element space
+ */
+class PetscPreconditionerSpaceDependent : public PetscPCSolver {
+protected:
+  /// @brief Finite element space defining the operator
+  mfem::ParFiniteElementSpace* fespace_ = nullptr;
+
+public:
+  /**
+   * @brief Construct a PetscPreconditionerSpaceDependent without an explicit operator (must be set later with
+   * SetOperator())
+   * @param comm The MPI communicator used by the vectors and matrices in the solve
+   * @param pc_type The PETSc PCType of the preconditioner
+   * @param prefix Prefix for all PETSc command line options options
+   *
+   * @note Additional arguments for the preconditioner can be set via the command line, e.g. `-pc_jacobi_type rowsum`.
+   * Use `-pc_type TYPE -help` for a complete list of options.
+   * @note If @a prefix is provided, the options should be set as `-[prefix]_[option] [value]`.
+   */
+  PetscPreconditionerSpaceDependent(MPI_Comm& comm, PCType pc_type = PCHMG, const std::string& prefix = std::string())
+      : PetscPCSolver(comm, pc_type, prefix)
+  {
+  }
+
+  /**
+   * @brief Construct a PetscPreconditionerSpaceDependent to approximate A^{-1}
+   * @param A The mfem::PetscParMatrix for which the preconditioner is constructed
+   * @param pc_type The PETSc PCType of the preconditioner
+   * @param prefix Prefix for all PETSc command line options options
+   *
+   * @note Additional arguments for the preconditioner can be set via the command line, e.g. `-pc_jacobi_type rowsum`.
+   * Use `-pc_type TYPE -help` for a complete list of options.
+   * @note If @a prefix is provided, the options should be set as `-[prefix]_[option] [value]`.
+   */
+  PetscPreconditionerSpaceDependent(mfem::PetscParMatrix& A, PCType pc_type = PCHMG,
+                                    const std::string& prefix = std::string())
+      : PetscPCSolver(A, pc_type, prefix)
+  {
+  }
+
+  /**
+   * @brief Construct a PetscPreconditionerSpaceDependent from a generic operator
+   * @param comm The MPI communicator used by the vectors and matrices in the solve
+   * @param op mfem::Operator used to construct the preconditioner
+   * @param pc_type The PETSc PCType of the preconditioner
+   * @param prefix Prefix for all PETSc command line options options
+   *
+   * @note Additional arguments for the preconditioner can be set via the command line, e.g. `-pc_jacobi_type rowsum`.
```

Comment:
Was this tested with the driver? I am concerned that the CLI11 command line parsing might mess this up.

Reply:
Ah, we'll have to use `allow_extras()` on the CLI11 app. Luckily, none of the PETSc options are valid CLI11 flags (they only use one `-`), so there shouldn't ever be conflicts.